pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -285,16 +299,29 @@ func (h *HistogramFunctionOverInstantVector) saveFloatsToGroup(fPoints []promql.
 			return err
 		}
 		g.pointBuckets = g.pointBuckets[:h.timeRange.StepCount]
+
+		g.pointBucketsIndex, err = types.UintSlicePool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+		g.pointBucketsIndex = g.pointBucketsIndex[:h.timeRange.StepCount]
 	}
 	for _, f := range fPoints {
 		pointIdx := h.timeRange.PointIndex(f.T)
-		g.pointBuckets[pointIdx] = append(
-			g.pointBuckets[pointIdx],
-			bucket{
-				upperBound: upperBound,
-				count:      f.F,
-			},
-		)
+
+		if g.pointBuckets[pointIdx] == nil {
+			// Remaining series count + 1 since we decrement the series count early to simplify each return point.
+			maxBuckets := int(g.remainingSeriesCount) + 1
+			g.pointBuckets[pointIdx], err = bucketSliceBucketedPool.Get(maxBuckets, h.memoryConsumptionTracker)
+			if err != nil {
+				return err
+			}
+			g.pointBuckets[pointIdx] = g.pointBuckets[pointIdx][:maxBuckets]
+		}
+
+		g.pointBuckets[pointIdx][g.pointBucketsIndex[pointIdx]].upperBound = upperBound
+		g.pointBuckets[pointIdx][g.pointBucketsIndex[pointIdx]].count = f.F
+		g.pointBucketsIndex[pointIdx]++
```

Comment:
Do we need `pointBucketsIndex`? Couldn't we leave `pointBuckets` with length 0 (rather than reslicing to length `maxBuckets` on line 319 above), and append to it here? This will use the existing capacity of the slice.

Reply:
This allows us to reuse Bucket{} without creating a new object. Hence we can set upperBound directly. We keep track of the index so that we re-slice off the unused buckets before the quantile calculation

Reply:
`pointBuckets` is defined as `[]buckets`, so each element is stored by value, not as a pointer, so there should be no difference.